### PR TITLE
Adding Windows support for InPlace Pod Vertical Scaling

### PR DESCRIPTION
**Pull request by fabi200123**
#### What type of PR is this?
/kind feature
/sig Windows

#### What this PR does / why we need it:
It adds Windows support for the In-place Pod Vertical Scaling feature.

#### Does this PR introduce a user-facing change?
```release-note
Added Windows support for InPlace Pod Vertical Scaling feature. 
```

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
[FEATURE]: https://github.com/kubernetes/kubernetes/pull/102884

/cc @claudiubelu

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
- KEP: https://github.com/kubernetes/enhancements/issues/1287
/ok-to-test
/triage accepted
/priority important-longterm

@fabi200123 please include the release notes section to the pr description

**Comment by salaxander** [comment]
Hey @fabi200123 @marosset - 1.26 bug triage team checking in! Just wanted to see if we're on track to get this in for the 1.26 release?

Thanks!

**Comment by vinaykul** [comment]
@fabi200123 Thanks for working on adding Windows support for In-Place Resize. Since this can be directly targeted to main branch, can I close [your PR](https://github.com/vinaykul/kubernetes/pull/16) targeting my mothership [PR 102884](https://github.com/kubernetes/kubernetes/pull/102884)?

**Comment by fabi200123** [comment]
Yes, you can close this!

**Comment by marosset** [comment]
It looks like https://github.com/vinaykul/kubernetes/pull/16 is closed.
Should we cherry-pick this and include it in https://github.com/kubernetes/kubernetes/pull/102884?

**Comment by matthyx** [comment]
@fabi200123 can you rebase and add a proper release note?

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by valaparthvi** [comment]
Hi there @fabi200123! I am from the Bug triage release team for 1.27. I am checking in to see if this PR is set to go in 1.27.

**Comment by valaparthvi** [comment]
Hi there @fabi200123! I am from the Bug triage release team for 1.27. I am checking in to see if this PR is set to go in 1.27.

**Comment by valaparthvi** [comment]
> Hi there @fabi200123! I am from the Bug triage release team for 1.27. I am checking in to see if this PR is set to go in 1.27.

Hi @fabi200123, @claudiubelu. With Code Freeze coming in on March 15, please ensure the PR goes in before that if you would like for it to be a part of v1.27 release. Post Code Freeze, we will only be able to accept Release Blockers and Critical issues.

**Comment by dims** [comment]
@fabi200123  please rebase!

File: pkg/kubelet/kuberuntime/kuberuntime_container_test.go
Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_container_test.go:266]: What is this test hoping to achieve? The current implementation for windows returns nil.

https://github.com/kubernetes/kubernetes/blob/20fc8a939c3394bd31edd06a893b75d1c312e27c/pkg/kubelet/kuberuntime/kuberuntime_container_windows.go#L136-L139

Should we not add functional code there and then add tests? Also, in practice we will have either `LinuxContainerResources` or `WindowsContainerResources`, not both.

Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_container_test.go:266]: In practice, we won't have both `LinuxContainerResources` and `WindowsContainerResources` would we? Perhaps we should populate the input.Resources based on the platform the test is executing on and select either LCR or WCR?

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows_test.go
Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_container_windows_test.go:180]: Does it make sense to add a test case requesting zero CPU limit for windows?

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (vinaykul) [inline pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:94]: Do we ever enter this block?

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows_test.go
Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows_test.go:180]: I think so, I have added it

**Comment by bart0sh** [comment]
@fabi200123 please update release notes so that `do-not-merge/release-note-label-needed` removed, thanks.

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (jsturtevant) [inline pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: this formula was simplified on line 140 above in `calculateCPUMaximum`.  Can we update it to match to avoid confusion later?

I am having a hard time from just this code (with little context to be fair) as to what we are doing here.  looks like we are doing something similar but not quite the same as `calculateCPUMaximum`?

Comment (jsturtevant) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: can we re-use the function or is there subtle differences between the two?

Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: There is no cpuRequest for Windows, which is why I could not reuse that function

Comment (jsturtevant) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: oh I see this is the reverse of that function.  Can we update the formula in the comment on this line as it doesn't match what is in `calculateWindowsResources`?

I think this was what initially had me confused.  Bonus points for explicitly mentioning this is the reverse of that function.

File: pkg/kubelet/kuberuntime/kuberuntime_container_test.go
Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_test.go:654]: Reverted the change

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: Modified the function to match that one: `cpuLimitValue := runtimeStatusResources.CpuMaximum * int64(winstats.ProcessorCount()) / 10`

File: pkg/kubelet/kuberuntime/kuberuntime_container_test.go
Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_test.go]: The part with `if wc.Resources.CpuMaximum > 0' was introduced here [#97141](https://github.com/kubernetes/kubernetes/pull/97141). Before that, the `!isolatedByHyperv` was created in [#85856](https://github.com/kubernetes/kubernetes/pull/85856) based on this issue [#84804](https://github.com/kubernetes/kubernetes/issues/84804)

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:94]: The part with `if wc.Resources.CpuMaximum > 0' was introduced here [#97141](https://github.com/kubernetes/kubernetes/pull/97141). Before that, the `!isolatedByHyperv` was created in [#85856](https://github.com/kubernetes/kubernetes/pull/85856) based on this issue [#84804](https://github.com/kubernetes/kubernetes/issues/84804)

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (jsturtevant) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:158]: Can you update the comment? the comment doesn't match the current code and it is confusing

Comment (fabi200123) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go]: Sure, changed it to point to the calculateCPUMaximum function

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (marosset) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:94]: We should untangle this as a separate change IMO

**Comment by marosset** [comment]
/lgtm
@claudiubelu can you reivew too?
Thanks!

**Comment by valaparthvi** [comment]
Hello,
It's me again, Parthvi from the Bug Triage Release team 1.28 👋. Since this PR is a spill over from the last milestone, I wanted to check in early this time and make sure it's still on track for 1.28. Please let me know if there's any help you require from my end :)

Thanks.

File: pkg/kubelet/kuberuntime/kuberuntime_container_windows.go
Comment (vinaykul) [reply pkg/kubelet/kuberuntime/kuberuntime_container_windows.go:94]: SG

**Comment by vinaykul** [comment]
@fabi200123 @marosset Is there a plan to add / adapt pod resize E2E tests for Windows (create a CI job for it) ?

**Comment by marosset** [comment]
> @fabi200123 @marosset Is there a plan to add / adapt pod resize E2E tests for Windows (create a CI job for it) ?

@jsturtevant has been setting up a test pass for testing alpha features on Windows nodes. We ca use that to run pod resize tests!

**Comment by jsturtevant** [comment]
do we want to run it against this PR? Can you give me a pointer to the tests we want to run and the correct featuregate?

**Comment by vinaykul** [comment]
> do we want to run it against this PR? Can you give me a pointer to the tests we want to run and the correct featuregate?

Please see test/e2e/node/pod_resize.go 

This test makes assumptions about cgroups in order to verify that resize was actuated in containerd versions before 1.6.9 (containerd 1.6.6 in COS version I last saw). If you have a newer containerd, you can make small changes to the test to skip verifying cgroup values (or replace with Windows equivalent) and rely on pod status values.

**Comment by jsturtevant** [comment]
@fabi200123 can you start looking at the changes needed to skip the tests prior to 1.6.9? I don't think we need to verify that for windows as we generally recommend staying farther ahead on the contianerd version. 

@marosset @dcantah do we know what version of containerd is required for this support?

I will getting an alpha job configured to run the tests for `[Feature:InPlacePodVerticalScaling]`

**Comment by jsturtevant** [comment]
this is the job we can use to run the tests with label `[Feature:InPlacePodVerticalScaling]`. It also runs a few other tests to make sure cluster is working generally.  We can consolidate the job with other alpha features once it is merged.

/test pull-kubernetes-e2e-capz-windows-alpha-feature-vertical-pod-autoscaler

**Comment by jsturtevant** [comment]
Something went wrong, going to run one more time 
/test pull-kubernetes-e2e-capz-windows-alpha-feature-vertical-pod-autoscaler

**Comment by jsturtevant** [comment]
logs weren't being collected but should be now 
/test pull-kubernetes-e2e-capz-windows-alpha-feature-vertical-pod-autoscaler

**Comment by jsturtevant** [comment]
@fabi200123 the tests are running now. The set of Windows specific tests passed but the InPlace Pod Vertical Scaling tests are failing to start the containers. I don't see any logs for the pod that is restarting but do see this in kubelet log:

```
E0627 19:55:49.376297     704 pod_workers.go:1300] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"app\" with CrashLoopBackOff: \"back-off 10s restarting failed container=app pod=img-pull-registry.k8s.io-e2e-test-images-busybox-1.29-4-pkm7g_img-puller-694(da759be0-4d86-4ef2-9ab4-2dbcc0258286)\"" pod="img-puller-694/img-pull-registry.k8s.io-e2e-test-images-busybox-1.29-4-pkm7g" podUID="da759be0-4d86-4ef2-9ab4-2dbcc0258286"
```

**Comment by valaparthvi** [comment]
Hello @fabi200123,
I am checking in on behalf of the Bug Triage release team. We are only a week away from Code Freeze ([01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)). Please ensure your PR goes in by then if you are targeting 1.28 release. 

Please let me know if I can be of any assistance :)

**Comment by marosset** [comment]
/test pull-kubernetes-e2e-capz-windows-alpha-feature-vpa
Hopefully this will pass because https://github.com/kubernetes-sigs/windows-testing/pull/406 merged

**Comment by marosset** [comment]
OK - I think the CI issues where e2e.test wasn't getting built from the PR branch for the windows presubmit test passes have been resolved and this is ready for review!

**Comment by marosset** [comment]
/remove-sig api-machinery apps auth cli cloud-provider cluster-lifecycle contributor-experience network release security storage testing

**Comment by marosset** [comment]
/remove-ara apiserver code-generation dependency ipvs kubectl provider/gcp release-eng

**Comment by marosset** [comment]
/remove-area apiserver code-generation dependency ipvs kubectl provider/gcp release-eng

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
